File: scripts/diff_preprocessor.py
"""
Pre-processes a git diff of discovery JSON documents into a clean, structured
representation per API — stripping noise and surfacing semantically meaningful
additions, removals, and modifications.

The output feeds directly into the LLM prompt.
"""

import json
import logging
import subprocess
from typing import Optional

logger = logging.getLogger(__name__)

# Keys that change on every sync and carry no semantic value for developers
NOISE_KEYS = frozenset(
    [
        "revision",
        "etag",
        "rootUrl",
        "servicePath",
        "batchPath",
        "baseUrl",
        "basePath",
        "documentationLink",
        "ownerDomain",
        "ownerName",
        "packagePath",
        "id",  # top-level only — filtered by path depth check
    ]
)

# Top-level path segments that are structural/infrastructure, not developer API surface
# e.g. endpoints[] changes are just regional URL shuffles
NOISE_PATH_PREFIXES = frozenset(["endpoints"])

# Keys that are only descriptive, not structural
DESCRIPTION_ONLY_KEYS = frozenset(["description", "title"])

# Maximum number of change entries per category to send to the LLM
MAX_ENTRIES_PER_CATEGORY = 60


def get_changed_discovery_files(base_ref: str, head_ref: str) -> list[str]:
    """Return list of discovery JSON filenames that changed between two refs."""
    result = subprocess.run(
        ["git", "diff", "--name-only", base_ref, head_ref, "--", "discoveries/"],
        capture_output=True,
        text=True,
        check=False,
    )
    if result.returncode != 0:
        logger.warning(f"git diff --name-only failed: {result.stderr}")
        return []
    paths = [
        p.strip() for p in result.stdout.splitlines() if p.strip().endswith(".json")
    ]
    return paths  # e.g. ['discoveries/bigquery.v2.json', ...]


def get_file_content_at_ref(ref: str, filepath: str) -> str:
    """Return full file content at a specific git ref, or '' if it didn't exist."""
    result = subprocess.run(
        ["git", "show", f"{ref}:{filepath}"],
        capture_output=True,
        text=True,
        check=False,
    )
    if result.returncode != 0:
        return ""  # file didn't exist at this ref (new or deleted)
    return result.stdout


def _flatten_json(obj, prefix: str = "") -> dict[str, object]:
    """Recursively flatten a JSON object into dot-notation paths."""
    items: dict[str, object] = {}
    if isinstance(obj, dict):
        for k, v in obj.items():
            new_key = f"{prefix}.{k}" if prefix else k
            if isinstance(v, (dict, list)):
                items.update(_flatten_json(v, new_key))
            else:
                items[new_key] = v
    elif isinstance(obj, list):
        for i, v in enumerate(obj):
            new_key = f"{prefix}[{i}]"
            if isinstance(v, (dict, list)):
                items.update(_flatten_json(v, new_key))
            else:
                items[new_key] = v
    return items


def _is_noise_path(path: str) -> bool:
    """Return True if a change at this path is pure noise (no developer impact)."""
    parts = path.split(".")
    # Top-level noise keys (e.g. revision, etag)
    if parts[0] in NOISE_KEYS:
        return True
    # Top-level structural prefixes (e.g. endpoints[N].location — infra noise)
    top_segment = parts[0].split("[")[0]
    if top_segment in NOISE_PATH_PREFIXES:
        return True
    # The key at the leaf level is a noise key
    leaf = parts[-1].split("[")[0]  # strip array index
    if leaf in NOISE_KEYS:
        return True
    return False


def _is_description_only(path: str) -> bool:
    leaf = path.split(".")[-1].split("[")[0]
    return leaf in DESCRIPTION_ONLY_KEYS


def detect_breaking_changes(
    old_flat: dict[str, object],
    new_flat: dict[str, object],
    added_paths: list[str],
    removed_paths: list[str],
    modified_paths: list[str],
) -> tuple[bool, list[str]]:
    """Deterministically evaluates whether changes constitute backward-incompatible API changes.

    Follows Google AIP-180 rules:
    - Removal of an active method / RPC
    - Removal of a parameter from an active method
    - Changing an optional parameter to required (required: true)
    - Removal of an existing schema property
    - Changing property data type (type or $ref)
    - Making a mutable schema property read-only / immutable
    - Changing HTTP verb or URI path template of an existing method

    Returns:
        (is_breaking: bool, reasons: list[str])
    """
    reasons: list[str] = []

    # 1. Inspect Added Paths (New required parameters or readOnly constraints on existing resources)
    for p in added_paths:
        parts = p.split(".")
        new_val = new_flat.get(p)
        if (
            parts[-1] == "required"
            and "parameters" in parts
            and new_val in (True, "true")
        ):
            param_name = parts[-2] if len(parts) >= 2 else "parameter"
            reason = f"Parameter '{param_name}' was changed to strictly required"
            if reason not in reasons:
                reasons.append(reason)
        elif (
            parts[-1] == "readOnly"
            and "schemas" in parts
            and "properties" in parts
            and new_val in (True, "true")
        ):
            s_idx = parts.index("schemas")
            p_idx = parts.index("properties")
            schema_name = parts[s_idx + 1] if len(parts) > s_idx + 1 else "Schema"
            prop_name = parts[p_idx + 1] if len(parts) > p_idx + 1 else "property"
            reason = f"Property '{prop_name}' in schema '{schema_name}' was made read-only / immutable"
            if reason not in reasons:
                reasons.append(reason)

    # 2. Inspect Removed Paths
    for p in removed_paths:
        parts = p.split(".")

        # Removed API method
        if "methods" in parts:
            idx = parts.index("methods")
            if len(parts) > idx + 1:
                method_name = parts[idx + 1]
                if len(parts) == idx + 2 or (
                    len(parts) == idx + 3
                    and parts[idx + 2] in ("httpMethod", "id", "path")
                ):
                    reason = f"Removed API method '{method_name}'"
                    if reason not in reasons:
                        reasons.append(reason)

        # Removed parameter from method
        if "parameters" in parts and "methods" in parts:
            idx = parts.index("parameters")
            if len(parts) > idx + 1:
                param_name = parts[idx + 1]
                if len(parts) == idx + 2 or (
                    len(parts) == idx + 3
                    and parts[idx + 2] in ("type", "location", "format")
                ):
                    reason = f"Removed parameter '{param_name}' from method"
                    if reason not in reasons:
                        reasons.append(reason)

        # Removed schema property
        if "schemas" in parts and "properties" in parts:
            s_idx = parts.index("schemas")
            p_idx = parts.index("properties")
            if s_idx < p_idx and len(parts) > p_idx + 1:
                schema_name = parts[s_idx + 1]
                prop_name = parts[p_idx + 1]
                if len(parts) == p_idx + 2 or (
                    len(parts) == p_idx + 3
                    and parts[p_idx + 2] in ("type", "$ref", "format")
                ):
                    reason = (
                        f"Removed property '{prop_name}' from schema '{schema_name}'"
                    )
                    if reason not in reasons:
                        reasons.append(reason)

    # 3. Inspect Modified Paths
    for p in modified_paths:
        parts = p.split(".")
        old_val = old_flat.get(p)
        new_val = new_flat.get(p)

        # Parameter made required
        if parts[-1] == "required" and "parameters" in parts:
            if old_val in (False, None, "false") and new_val in (True, "true"):
                param_name = parts[-2] if len(parts) >= 2 else "parameter"
                reason = f"Parameter '{param_name}' was changed to strictly required"
                if reason not in reasons:
                    reasons.append(reason)

        # Schema property type, $ref, or readOnly changes
        if "schemas" in parts and "properties" in parts:
            s_idx = parts.index("schemas")
            p_idx = parts.index("properties")
            if s_idx < p_idx and len(parts) > p_idx + 2:
                schema_name = parts[s_idx + 1]
                prop_name = parts[p_idx + 1]
                attr = parts[p_idx + 2]

                if attr == "type" and old_val != new_val:
                    reason = f"Property '{prop_name}' in schema '{schema_name}' changed type from '{old_val}' to '{new_val}'"
                    if reason not in reasons:
                        reasons.append(reason)
                elif attr == "$ref" and old_val != new_val:
                    reason = f"Property '{prop_name}' in schema '{schema_name}' changed referenced type from '{old_val}' to '{new_val}'"
                    if reason not in reasons:
                        reasons.append(reason)
                elif (
                    attr == "readOnly"
                    and (old_val in (False, None))
                    and (new_val is True)
                ):
                    reason = f"Property '{prop_name}' in schema '{schema_name}' was made read-only / immutable"
                    if reason not in reasons:
                        reasons.append(reason)

        # Method HTTP verb or URI path template change
        if "methods" in parts:
            m_idx = parts.index("methods")
            if len(parts) > m_idx + 2:
                method_name = parts[m_idx + 1]
                attr = parts[m_idx + 2]
                if attr == "httpMethod" and old_val != new_val:
                    reason = f"Method '{method_name}' changed HTTP verb from '{old_val}' to '{new_val}'"
                    if reason not in reasons:
                        reasons.append(reason)
                elif attr == "path" and old_val != new_val:
                    reason = f"Method '{method_name}' URI path template changed from '{old_val}' to '{new_val}'"
                    if reason not in reasons:
                        reasons.append(reason)

    is_breaking = len(reasons) > 0
    return is_breaking, reasons


def build_structured_diff(
    filename: str, old_json_str: str, new_json_str: str
) -> Optional[dict]:
    """Produce a structured diff dict for a single API file.

    Returns None if the diff is purely noise (nothing developer-relevant changed).
    """
    try:
        old = json.loads(old_json_str) if old_json_str.strip() else {}
        new = json.loads(new_json_str) if new_json_str.strip() else {}
    except json.JSONDecodeError as e:
        logger.warning(f"Could not parse JSON for {filename}: {e}")
        return None

    old_flat = _flatten_json(old)
    new_flat = _flatten_json(new)

    old_keys = set(old_flat.keys())
    new_keys = set(new_flat.keys())

    added_paths = new_keys - old_keys
    removed_paths = old_keys - new_keys
    common_paths = old_keys & new_keys
    modified_paths = {p for p in common_paths if old_flat[p] != new_flat[p]}

    def _filter(paths):
        return [p for p in paths if not _is_noise_path(p)]

    added = _filter(added_paths)
    removed = _filter(removed_paths)
    modified = _filter(modified_paths)

    # Safeguard against document sweeps / transient whole-file deletions:
    # If the file was deleted (new is empty) or virtually all content was stripped without additions,
    # skip generating an update. True GCP API deprecations occur inside active discovery docs
    # with explicit "deprecated: true" flags.
    if (not new_flat and old_flat) or (
        len(new_flat) < 5 and len(removed) > 40 and not added
    ):
        logger.info(
            f"  {filename}: document wiped or deleted ({len(removed)} removals, {len(new_flat)} remaining). "
            "Skipping feed generation to prevent transient sweep false positives."
        )
        return None

    if not added and not removed and not modified:
        logger.info(f"  {filename}: only noise changes, skipping")
        return None

    # Compute deterministic breaking change analysis
    is_breaking, breaking_reasons = detect_breaking_changes(
        old_flat, new_flat, added, removed, modified
    )

    # Build structured entries (cap size to stay within token budget)
    def _added_entries(paths):
        out = []
        for p in sorted(paths)[:MAX_ENTRIES_PER_CATEGORY]:
            entry = {"path": p}
            val = new_flat[p]
            if not isinstance(val, (dict, list)):
                entry["value"] = val
            out.append(entry)
        return out

    def _removed_entries(paths):
        out = []
        for p in sorted(paths)[:MAX_ENTRIES_PER_CATEGORY]:
            entry = {"path": p}
            val = old_flat[p]
            if not isinstance(val, (dict, list)):
                entry["old_value"] = val
            out.append(entry)
        return out

    def _modified_entries(paths):
        out = []
        for p in sorted(paths)[:MAX_ENTRIES_PER_CATEGORY]:
            out.append({"path": p, "old": old_flat[p], "new": new_flat[p]})
        return out

    api_name = filename.replace(".json", "")  # e.g. bigquery.v2
    return {
        "api": api_name,
        "is_breaking": is_breaking,
        "breaking_reasons": breaking_reasons,
        "added": _added_entries(added),
        "removed": _removed_entries(removed),
        "modified": _modified_entries(modified),
        # Counts useful for prompt context
        "_stats": {
            "added_count": len(added),
            "removed_count": len(removed),
            "modified_count": len(modified),
            "description_only_modified": sum(
                1 for p in modified if _is_description_only(p)
            ),
            "is_breaking": is_breaking,
            "breaking_reasons": breaking_reasons,
        },
    }


def extract_structured_diffs(
    base_ref: str = "HEAD~1", head_ref: str = "HEAD"
) -> list[dict]:
    """Main entry point: returns a list of structured diffs, one per changed API.

    Args:
        base_ref: Git ref for the base (old) state.
        head_ref: Git ref for the head (new) state.

    Returns:
        List of structured diff dicts, one per API, filtered of noise-only changes.
    """
    logger.info(
        f"Detecting changed discovery files between {base_ref} and {head_ref} ..."
    )
    changed_paths = get_changed_discovery_files(base_ref, head_ref)

    if not changed_paths:
        logger.info("No discovery files changed — nothing to analyse")
        return []

    # Filter out the index file — it's not per-API content
    changed_paths = [p for p in changed_paths if not p.endswith("index.json")]
    logger.info(f"Found {len(changed_paths)} changed API discovery files")

    results = []
    for filepath in sorted(changed_paths):
        filename = filepath.split("/")[-1]  # e.g. bigquery.v2.json
        logger.info(f"Processing {filename} ...")
        old_content = get_file_content_at_ref(base_ref, filepath)
        new_content = get_file_content_at_ref(head_ref, filepath)
        diff = build_structured_diff(filename, old_content, new_content)
        if diff is not None:
            results.append(diff)

    logger.info(f"Produced {len(results)} non-trivial structured diffs")
    return results

File: scripts/diff_preprocessor.ts
/**
 * Pre-processes a git diff of discovery JSON documents into a clean, structured
 * representation per API — stripping noise and surfacing semantically meaningful
 * additions, removals, and modifications.
 *
 * The output feeds directly into the LLM prompt.
 */

import { execFileSync } from 'node:child_process';
import fs from 'node:fs';

// Keys that change on every sync and carry no semantic value for developers
export const NOISE_KEYS = new Set([
  'revision',
  'etag',
  'rootUrl',
  'servicePath',
  'batchPath',
  'baseUrl',
  'basePath',
  'documentationLink',
  'ownerDomain',
  'ownerName',
  'packagePath',
  'id', // top-level only — filtered by path depth check
]);

// Top-level path segments that are structural/infrastructure, not developer API surface
export const NOISE_PATH_PREFIXES = new Set(['endpoints']);

// Keys that are only descriptive, not structural
export const DESCRIPTION_ONLY_KEYS = new Set(['description', 'title']);

// Maximum number of change entries per category to send to the LLM
export const MAX_ENTRIES_PER_CATEGORY = 60;

export function getChangedDiscoveryFiles(baseRef: string, headRef = 'WORKTREE'): string[] {
  let args: string[];
  if (!headRef || headRef === 'WORKTREE') {
    args = ['diff', '--name-only', baseRef, '--', 'discoveries/'];
  } else {
    args = ['diff', '--name-only', baseRef, headRef, '--', 'discoveries/'];
  }

  try {
    const stdout = execFileSync('git', args, { encoding: 'utf-8', stdio: ['pipe', 'pipe', 'ignore'] });
    return stdout
      .split('\n')
      .map((p) => p.trim())
      .filter((p) => p.endsWith('.json'));
  } catch (err: any) {
    console.warn(`git diff --name-only failed: ${err.message}`);
    return [];
  }
}

export function getFileContentAtRef(ref: string, filepath: string): string {
  if (!ref || ref === 'WORKTREE') {
    try {
      return fs.readFileSync(filepath, 'utf-8');
    } catch {
      return '';
    }
  }

  try {
    return execFileSync('git', ['show', `${ref}:${filepath}`], {
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'ignore'],
    });
  } catch {
    return '';
  }
}

export function flattenJson(obj: any, prefix = ''): Record<string, any> {
  const items: Record<string, any> = {};
  if (obj !== null && typeof obj === 'object' && !Array.isArray(obj)) {
    for (const [k, v] of Object.entries(obj)) {
      const newKey = prefix ? `${prefix}.${k}` : k;
      if (v !== null && typeof v === 'object') {
        Object.assign(items, flattenJson(v, newKey));
      } else {
        items[newKey] = v;
      }
    }
  } else if (Array.isArray(obj)) {
    for (let i = 0; i < obj.length; i++) {
      const v = obj[i];
      const newKey = `${prefix}[${i}]`;
      if (v !== null && typeof v === 'object') {
        Object.assign(items, flattenJson(v, newKey));
      } else {
        items[newKey] = v;
      }
    }
  }
  return items;
}

export function isNoisePath(pathStr: string): boolean {
  const parts = pathStr.split('.');
  // Top-level noise keys (e.g. revision, etag)
  if (NOISE_KEYS.has(parts[0])) {
    return true;
  }
  // Top-level structural prefixes (e.g. endpoints[N].location — infra noise)
  const topSegment = parts[0].split('[')[0];
  if (NOISE_PATH_PREFIXES.has(topSegment)) {
    return true;
  }
  // The key at the leaf level is a noise key
  const leaf = parts[parts.length - 1].split('[')[0];
  if (NOISE_KEYS.has(leaf)) {
    return true;
  }
  return false;
}

export function isDescriptionOnly(pathStr: string): boolean {
  const parts = pathStr.split('.');
  const leaf = parts[parts.length - 1].split('[')[0];
  return DESCRIPTION_ONLY_KEYS.has(leaf);
}

export function detectBreakingChanges(
  oldFlat: Record<string, any>,
  newFlat: Record<string, any>,
  addedPaths: string[],
  removedPaths: string[],
  modifiedPaths: string[]
): [boolean, string[]] {
  const reasons: string[] = [];

  // 1. Inspect Added Paths (New required parameters or readOnly constraints on existing resources)
  for (const p of addedPaths) {
    const parts = p.split('.');
    const newVal = newFlat[p];
    if (
      parts[parts.length - 1] === 'required' &&
      parts.includes('parameters') &&
      (newVal === true || newVal === 'true')
    ) {
      const paramName = parts.length >= 2 ? parts[parts.length - 2] : 'parameter';
      const reason = `Parameter '${paramName}' was changed to strictly required`;
      if (!reasons.includes(reason)) reasons.push(reason);
    } else if (
      parts[parts.length - 1] === 'readOnly' &&
      parts.includes('schemas') &&
      parts.includes('properties') &&
      (newVal === true || newVal === 'true')
    ) {
      const sIdx = parts.indexOf('schemas');
      const pIdx = parts.indexOf('properties');
      const schemaName = parts.length > sIdx + 1 ? parts[sIdx + 1] : 'Schema';
      const propName = parts.length > pIdx + 1 ? parts[pIdx + 1] : 'property';
      const reason = `Property '${propName}' in schema '${schemaName}' was made read-only / immutable`;
      if (!reasons.includes(reason)) reasons.push(reason);
    }
  }

  // 2. Inspect Removed Paths
  for (const p of removedPaths) {
    const parts = p.split('.');

    // Removed API method
    if (parts.includes('methods')) {
      const idx = parts.indexOf('methods');
      if (parts.length > idx + 1) {
        const methodName = parts[idx + 1];
        if (
          parts.length === idx + 2 ||
          (parts.length === idx + 3 && ['httpMethod', 'id', 'path'].includes(parts[idx + 2]))
        ) {
          const reason = `Removed API method '${methodName}'`;
          if (!reasons.includes(reason)) reasons.push(reason);
        }
      }
    }

    // Removed parameter from method
    if (parts.includes('parameters') && parts.includes('methods')) {
      const idx = parts.indexOf('parameters');
      if (parts.length > idx + 1) {
        const paramName = parts[idx + 1];
        if (
          parts.length === idx + 2 ||
          (parts.length === idx + 3 && ['type', 'location', 'format'].includes(parts[idx + 2]))
        ) {
          const reason = `Removed parameter '${paramName}' from method`;
          if (!reasons.includes(reason)) reasons.push(reason);
        }
      }
    }

    // Removed default value from method parameter
    if (parts[parts.length - 1] === 'default' && parts.includes('parameters') && !parts.includes('schemas')) {
      const pIdx = parts.indexOf('parameters');
      if (parts.length > pIdx + 1) {
        const paramName = parts[pIdx + 1];
        const paramPrefix = parts.slice(0, pIdx + 2).join('.');
        if (Object.keys(newFlat).some((k) => k === paramPrefix || k.startsWith(`${paramPrefix}.`))) {
          const reason = `Default value for parameter '${paramName}' was removed`;
          if (!reasons.includes(reason)) reasons.push(reason);
        }
      }
    }

    // Removed schema property
    if (parts.includes('schemas') && parts.includes('properties')) {
      const sIdx = parts.indexOf('schemas');
      const pIdx = parts.indexOf('properties');
      if (sIdx < pIdx && parts.length > pIdx + 1) {
        const schemaName = parts[sIdx + 1];
        const propName = parts[pIdx + 1];
        if (
          parts.length === pIdx + 2 ||
          (parts.length === pIdx + 3 && ['type', '$ref', 'format'].includes(parts[pIdx + 2]))
        ) {
          const reason = `Removed property '${propName}' from schema '${schemaName}'`;
          if (!reasons.includes(reason)) reasons.push(reason);
        }
      }
    }
  }

  // 3. Inspect Modified Paths
  for (const p of modifiedPaths) {
    const parts = p.split('.');
    const oldVal = oldFlat[p];
    const newVal = newFlat[p];

    // Parameter made required
    if (parts[parts.length - 1] === 'required' && parts.includes('parameters')) {
      if ((oldVal === false || oldVal === null || oldVal === 'false') && (newVal === true || newVal === 'true')) {
        const paramName = parts.length >= 2 ? parts[parts.length - 2] : 'parameter';
        const reason = `Parameter '${paramName}' was changed to strictly required`;
        if (!reasons.includes(reason)) reasons.push(reason);
      }
    }

    // Parameter default value changed
    if (parts[parts.length - 1] === 'default' && parts.includes('parameters') && !parts.includes('schemas')) {
      const pIdx = parts.indexOf('parameters');
      if (parts.length > pIdx + 1) {
        const paramName = parts[pIdx + 1];
        const reason = `Default value for parameter '${paramName}' changed from '${oldVal}' to '${newVal}'`;
        if (!reasons.includes(reason)) reasons.push(reason);
      }
    }

    // Schema property type, $ref, or readOnly changes
    if (parts.includes('schemas') && parts.includes('properties')) {
      const sIdx = parts.indexOf('schemas');
      const pIdx = parts.indexOf('properties');
      if (sIdx < pIdx && parts.length > pIdx + 2) {
        const schemaName = parts[sIdx + 1];
        const propName = parts[pIdx + 1];
        const attr = parts[pIdx + 2];

        if (attr === 'type' && oldVal !== newVal) {
          const reason = `Property '${propName}' in schema '${schemaName}' changed type from '${oldVal}' to '${newVal}'`;
          if (!reasons.includes(reason)) reasons.push(reason);
        } else if (attr === '$ref' && oldVal !== newVal) {
          const reason = `Property '${propName}' in schema '${schemaName}' changed referenced type from '${oldVal}' to '${newVal}'`;
          if (!reasons.includes(reason)) reasons.push(reason);
        } else if (attr === 'readOnly' && (oldVal === false || oldVal === null) && newVal === true) {
          const reason = `Property '${propName}' in schema '${schemaName}' was made read-only / immutable`;
          if (!reasons.includes(reason)) reasons.push(reason);
        }
      }
    }

    // Method HTTP verb or URI path template change
    if (parts.includes('methods')) {
      const mIdx = parts.indexOf('methods');
      if (parts.length > mIdx + 2) {
        const methodName = parts[mIdx + 1];
        const attr = parts[mIdx + 2];
        if (attr === 'httpMethod' && oldVal !== newVal) {
          const reason = `Method '${methodName}' changed HTTP verb from '${oldVal}' to '${newVal}'`;
          if (!reasons.includes(reason)) reasons.push(reason);
        } else if (attr === 'path' && oldVal !== newVal) {
          const reason = `Method '${methodName}' URI path template changed from '${oldVal}' to '${newVal}'`;
          if (!reasons.includes(reason)) reasons.push(reason);
        }
      }
    }
  }

  // 4. Parameter Enum Value Removal
  const enumParamPrefixes = new Set<string>();
  for (const k of Object.keys(oldFlat)) {
    if (k.includes('.enum[')) {
      const parts = k.split('.');
      if (parts.includes('parameters') && !parts.includes('schemas')) {
        const pIdx = parts.indexOf('parameters');
        if (parts.length > pIdx + 1) {
          const paramPrefix = parts.slice(0, pIdx + 2).join('.');
          enumParamPrefixes.add(paramPrefix);
        }
      }
    }
  }

  for (const paramPrefix of Array.from(enumParamPrefixes).sort()) {
    if (!Object.keys(newFlat).some((k) => k === paramPrefix || k.startsWith(`${paramPrefix}.`))) {
      continue; // Parameter itself was removed; handled by parameter removal check
    }

    const parts = paramPrefix.split('.');
    const pIdx = parts.indexOf('parameters');
    const paramName = parts[pIdx + 1];

    const oldEnums = new Set(
      Object.entries(oldFlat)
        .filter(([k]) => k.startsWith(`${paramPrefix}.enum[`))
        .map(([, v]) => String(v))
    );
    const newEnums = new Set(
      Object.entries(newFlat)
        .filter(([k]) => k.startsWith(`${paramPrefix}.enum[`))
        .map(([, v]) => String(v))
    );

    const removedEnums = Array.from(oldEnums).filter((x) => !newEnums.has(x));
    for (const val of removedEnums.sort()) {
      const reason = `Removed enum value '${val}' from parameter '${paramName}'`;
      if (!reasons.includes(reason)) reasons.push(reason);
    }
  }

  const isBreaking = reasons.length > 0;
  return [isBreaking, reasons];
}

export function extractMethodIdentifiersFromPaths(paths: Set<string> | string[]): string[] {
  const methods = new Set<string>();
  for (const p of paths) {
    const parts = p.split('.');
    if (parts.includes('methods')) {
      const idx = parts.indexOf('methods');
      if (parts.length > idx + 1) {
        const methodName = parts[idx + 1];
        const resParts = parts.slice(0, idx).filter((part) => part !== 'resources' && part !== 'methods');
        const methodId = resParts.length > 0 ? `${resParts.join('.')}.${methodName}` : methodName;
        methods.add(methodId.toLowerCase());
      }
    }
  }
  return Array.from(methods).sort();
}

export function buildStructuredDiff(
  filename: string,
  oldJsonStr: string,
  newJsonStr: string
): Record<string, any> | null {
  let oldObj: any = {};
  let newObj: any = {};

  try {
    oldObj = oldJsonStr && oldJsonStr.trim() ? JSON.parse(oldJsonStr) : {};
    newObj = newJsonStr && newJsonStr.trim() ? JSON.parse(newJsonStr) : {};
  } catch (err: any) {
    console.warn(`Could not parse JSON for ${filename}: ${err.message}`);
    return null;
  }

  const oldFlat = flattenJson(oldObj);
  const newFlat = flattenJson(newObj);

  const oldKeys = new Set(Object.keys(oldFlat));
  const newKeys = new Set(Object.keys(newFlat));

  const addedPaths = Array.from(newKeys).filter((k) => !oldKeys.has(k));
  const removedPaths = Array.from(oldKeys).filter((k) => !newKeys.has(k));
  const commonPaths = Array.from(oldKeys).filter((k) => newKeys.has(k));
  const modifiedPaths = commonPaths.filter((p) => oldFlat[p] !== newFlat[p]);

  const filterNoise = (paths: string[]) => paths.filter((p) => !isNoisePath(p));

  const added = filterNoise(addedPaths);
  const removed = filterNoise(removedPaths);
  const modified = filterNoise(modifiedPaths);

  if ((Object.keys(newFlat).length === 0 && Object.keys(oldFlat).length > 0) ||
      (Object.keys(newFlat).length < 5 && removed.length > 40 && added.length === 0)) {
    console.info(`  ${filename}: document wiped or deleted (${removed.length} removals). Skipping.`);
    return null;
  }

  if (added.length === 0 && removed.length === 0 && modified.length === 0) {
    return null;
  }

  const [isBreaking, breakingReasons] = detectBreakingChanges(
    oldFlat,
    newFlat,
    added,
    removed,
    modified
  );

  const addedEntries = added.sort().slice(0, MAX_ENTRIES_PER_CATEGORY).map((p) => {
    const entry: Record<string, any> = { path: p };
    const val = newFlat[p];
    if (val !== null && typeof val !== 'object') entry.value = val;
    return entry;
  });

  const removedEntries = removed.sort().slice(0, MAX_ENTRIES_PER_CATEGORY).map((p) => {
    const entry: Record<string, any> = { path: p };
    const val = oldFlat[p];
    if (val !== null && typeof val !== 'object') entry.old_value = val;
    return entry;
  });

  const modifiedEntries = modified.sort().slice(0, MAX_ENTRIES_PER_CATEGORY).map((p) => ({
    path: p,
    old: oldFlat[p],
    new: newFlat[p],
  }));

  const apiName = filename.replace('.json', '');
  const allChangedPaths = new Set([...added, ...removed, ...modified]);
  const extractedMethods = extractMethodIdentifiersFromPaths(allChangedPaths);
  const targetPaths = Array.from(allChangedPaths).sort();

  return {
    api: apiName,
    is_breaking: isBreaking,
    breaking_reasons: breakingReasons,
    added: addedEntries,
    removed: removedEntries,
    modified: modifiedEntries,
    extracted_methods: extractedMethods,
    target_paths: targetPaths,
    _stats: {
      added_count: added.length,
      removed_count: removed.length,
      modified_count: modified.length,
      description_only_modified: modified.filter(isDescriptionOnly).length,
      is_breaking: isBreaking,
      breaking_reasons: breakingReasons,
    },
  };
}

export function extractStructuredDiffs(
  baseRef = 'HEAD',
  headRef = 'WORKTREE'
): Array<Record<string, any>> {
  console.log(`Detecting changed discovery files between ${baseRef} and ${headRef} ...`);
  const changedPaths = getChangedDiscoveryFiles(baseRef, headRef).filter(
    (p) => !p.endsWith('index.json')
  );

  if (changedPaths.length === 0) {
    console.log('No discovery files changed — nothing to analyse');
    return [];
  }

  const results: Array<Record<string, any>> = [];
  for (const filepath of changedPaths.sort()) {
    const filename = filepath.split('/').pop() || filepath;
    const oldContent = getFileContentAtRef(baseRef, filepath);
    const newContent = getFileContentAtRef(headRef, filepath);
    const diff = buildStructuredDiff(filename, oldContent, newContent);
    if (diff !== null) {
      results.push(diff);
    }
  }

  console.log(`Produced ${results.length} non-trivial structured diffs`);
  return results;
}
